### PR TITLE
feat: add asset edit functionality

### LINKED ISSUE #10: [Feature] Provide "edit" asset function
## Motivation / Problem
Current state "Publishing":

```mermaid
classDiagram
    note for DDO "ocean.js spec DDO"
    AssetBuilder *-- NautilusAsset
    NautilusAsset *-- NautilusDDO
    NautilusDDO *-- DDO
    class NautilusAsset{
        +NautilusDDO nautilusDDO
    }
    class AssetBuilder{
        +NautilusAsset nautilusAsset
    }
    class NautilusDDO{
        +DDO ddo
    }
    class DDO{
        +object metadata
        +service[] services
    }
```

We want to be able to reuse publishing for editing functionality.

## Solution
- [x] Create constructors to allow pre-loading the DDO object of `NautilusDDO` with aquarius metadata
  - [x] NautilusDDO: `static createFromDDO(ddo: DDO)`
  - [x] NautilusAsset
  - [x] AssetBuilder
- [x] Helper function to create `NautilusDDO` / `NautilusAsset` from a given did
- [x] New `Nautilus.edit()` function that re-uses the existing publishing flow
  - [x] optional: if service changes, we need a new function to edit services (research ocean specs)
  - [x] optional: if new services, see [Step 2. `createDatatokenAndPricing`](https://github.com/deltaDAO/nautilus/blob/25cd87c98291027ad1d413d66d1a2dbca48e835b/src/Nautilus/Nautilus.ts#L136-L164)
  - [x] always: see [Step 3. `getDDO` and `publishDDO`](https://github.com/deltaDAO/nautilus/blob/25cd87c98291027ad1d413d66d1a2dbca48e835b/src/Nautilus/Nautilus.ts#L166-L179)
- [x] Tests if edits / updates are overwritten (applied) correctly
  - [x] Metadata changes
  - [x] Service changes
  - [x] ...
 - [ ] confirm test-edits with asset queries (hard because of aquarius delay, maybe we can trigger caching, see: https://docs.oceanprotocol.com/developers/aquarius/asset-requests#trigger-caching)

### PR DESCRIPTION
**Pull request by Abrom8**
closes #10 

## Proposed Changes
- Edit asset functionality

### CODE REVIEW

File: src/Nautilus/Asset/Service/ServiceBuilder.ts [hunk 1]
```diff
@@ -172,35 +172,41 @@ export class ServiceBuilder<
...
+  addTrustedAlgorithms(algorithms: PublisherTrustedAlgorithm[]) {
     if (this.service.type !== 'compute') {
       throw new Error('Illegal operation, asset is not a compute asset!')
     }
 
-    if (!this.service.compute.publisherTrustedAlgorithms) {
-      this.service.compute.publisherTrustedAlgorithms = [algorithm]
-      return this
+    // Check if the algorithms array is empty
+    if (algorithms.length === 0) {
```
Comment (moritzkirstein) [inline src/Nautilus/Asset/Service/ServiceBuilder.ts:181]: issue: could be undefined
```suggestion
    if (algorithms?.length === 0) {
```

File: src/utils/helpers/trusted-algorithms.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  PublisherTrustedAlgorithm,
+  getHash
+} from '@oceanprotocol/lib'
+import { Nautilus } from '../../Nautilus'
+import { getAsset } from '../aquarius'
+
+// TODO replace hardcoded service index 0 with service id once supported by the stack
+export async function getPublisherTrustedAlgorithms(
+  dids: string[],
+  nautilus: Nautilus
```
Comment (moritzkirstein) [inline src/utils/helpers/trusted-algorithms.ts:14]: suggestion: as this is a utility function, I prefer for it to work without relying on any repo class instances. In this case I think `nautilus` is only used to retrieve the `metadataCacheUri` from the config. To break responsibilities a bit better here, imo the helper should simply work with passing a `metadataCacheUri` parameter.

```ts
export async function getPublisherTrustedAlgorithms(
  dids: string[],
  metadataCacheUri: string
) {
  const assetPromises = dids.map((did) =>
    getAsset(metadataCacheUri, did)
  )
// ... 
}
```

File: src/utils/helpers/trusted-algorithms.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    const filesChecksum = await getFileDidInfo(
+      asset?.id,
+      asset?.services?.[0]?.id,
+      asset?.services?.[0]?.serviceEndpoint
+    )
+    if (!filesChecksum?.[0]?.checksum)
+      throw new Error(`Unable to get fileChecksum for asset ${asset.id}`)
+
+    const containerChecksum =
```
Comment (moritzkirstein) [inline src/utils/helpers/trusted-algorithms.ts:44]: suggestion: a comment linking to the definition/spec would be useful here

**Review by moritzkirstein — CHANGES_REQUESTED** [review]

File: src/Nautilus/Nautilus.ts [hunk 4]
```diff
@@ -186,6 +188,70 @@ export class Nautilus {
...
+              ...service.datatokenCreateParams,
+              minter: asset.owner,
+              paymentCollector: asset.owner
+            }
+          })
+
+          service.datatokenAddress = datatokenAddress
+
+          return { service, datatokenAddress, tx }
+        }
```
[the remark this answers is not in the document]
Comment (moritzkirstein) [reply src/Nautilus/Nautilus.ts:231]: comment: add a TODO to check if related problems can be solved on caching layer or not

File: src/index.ts [hunk 5]
```diff
@@ -3,3 +3,4 @@ export * from './access'
 export * from './compute'
 export * from './publish'
 export * from './Nautilus'
+export { getPublisherTrustedAlgorithms } from './utils/helpers/trusted-algorithms'
```
Comment (moritzkirstein) [inline src/index.ts:6]: reminder: can be removed
optionally: export * from utils can be added

File: src/Nautilus/Asset/NautilusDDO.ts [hunk 6]
```diff
@@ -68,20 +112,42 @@ export class NautilusDDO {
...
+        (service.editExistingService && service.filesEdited) ||
+        (service.editExistingService && service.serviceEndpointEdited) ||
+        service.pricing
+      )
 
-    // build new services if needed
-    const newServices = await this.buildDDOServices()
+      if (service.id && isFilesObjectChanged) {
+        this.removeServices.push(service.id)
+      }
```
[the remark this answers is not in the document]
Comment (Abrom8) [reply src/Nautilus/Asset/NautilusDDO.ts:108]: Created #61 for that.

File: src/utils/aquarius.ts [hunk 7]
```diff
@@ -25,5 +28,63 @@ export async function getAsset(
...
       LoggerInstance.error(error.message)
     }
+    throw error
+  }
+}
+
+export async function getAssets(
+  metadataCacheUri: string,
+  dids: string[]
+): Promise<{ [key: string]: Asset }> {
```
Comment (moritzkirstein) [inline src/utils/aquarius.ts:38]: nitpick: we could rename `key` to `did` to indicate to the user how these are used

File: src/utils/helpers/trusted-algorithms.ts [hunk 8]
```diff
@@ -71,3 +70,47 @@ async function getFileDidInfo(
     throw new Error(`[Initialize check file did] Error:' ${error}`)
   }
 }
+
+export async function resolvePublisherTrustedAlgorithms(
+  nautilusDDOServices: NautilusService<ServiceTypes, FileTypes>[],
+  metadataCacheUri: string
+) {
+  for (const service of nautilusDDOServices) {
+    if (service.addedPublisherTrustedAlgorithms.length > 0) {
```
Comment (moritzkirstein) [inline src/utils/helpers/trusted-algorithms.ts:79]: suggestion: it makes sense utilizing the early return / continue pattern here

File: src/utils/constants.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Asset, DDO } from '@oceanprotocol/lib'
+
+export type AssetSpecificProps = Exclude<keyof Asset, keyof DDO>
+
+// these are the props which extend the DDO interface to be an aquarius Asset
+// this const is needed to strip these away from an Asset to convert it into an DDO
+export const AQUARIUS_ASSET_EXTENDED_DDO_PROPS: AssetSpecificProps[] = [
```
Comment (moritzkirstein) [inline src/utils/constants.ts:7]: comment: let's add a TODO to look into type optimization at some point. Not a high prio for this PR imo

**Review by moritzkirstein — APPROVED** [review]

**Pull request merged**